libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,205 @@
+
+// Formula obtained from section 13.9 of the reference manual. Returns reading in kelvin
+static uint16_t prv_get_temp(uint16_t reading) {
+  uint16_t ts_cal1 = *(uint16_t*)TS_CAL1;
+  uint16_t ts_cal2 = *(uint16_t*)TS_CAL2;
+
+  reading = ((110 - 30) * (reading - ts_cal1)) / (ts_cal2 - ts_cal1) + 30;
+
+  return reading + 273;
+}
+
+// Formula obtained from section 13.9 of the reference manual. Returns Vdda in mV
+static uint16_t prv_get_vdda(uint16_t reading) {
+  uint16_t vrefint_cal = *(uint16_t*)VREFINT_CAL;
+  reading = (3300 * vrefint_cal) / reading;
+  return reading;
+}
+
+static StatusCode prv_channel_valid(ADCChannel adc_channel) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (!(ADC1->CHSELR & (1 << adc_channel))) {
+    return STATUS_CODE_EMPTY;
+  }
+
+  return STATUS_CODE_OK;
+}
+
+void adc_init(ADCMode adc_mode) {
+  ADC_DeInit(ADC1);
+
+  // Once the ADC has been reset, enable it with the given settings
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    .ADC_Resolution = ADC_Resolution_12b,
+    .ADC_ContinuousConvMode = adc_mode,
+    .ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None,
+    .ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_TRGO,
+    .ADC_DataAlign = ADC_DataAlign_Right,
+    .ADC_ScanDirection = ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  ADC_WaitModeCmd(ADC1, ENABLE);
+  ADC_AutoPowerOffCmd(ADC1, !adc_mode);
+
+  // Enable interrupts for the end of each conversion
+  stm32f0xx_interrupt_nvic_enable(ADC1_COMP_IRQn, INTERRUPT_PRIORITY_NORMAL);
+  ADC_ITConfig(ADC1, ADC_IER_EOCIE, ENABLE);
+  ADC_ITConfig(ADC1, ADC_IER_EOSEQIE, ENABLE);
+
+  // Initialize static varables
+  s_adc_status.continuous = adc_mode;
+  s_adc_status.sequence = 0;
+
+  if (adc_mode) {
+    ADC_StartOfConversion(ADC1);
+  }
+
+  // Configure internal reference channel to run by default for voltage conversions
+  adc_set_channel(ADC_CHANNEL_REF, ENABLE);
+}
+
+StatusCode adc_set_channel(ADCChannel adc_channel, bool new_state) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (new_state) {
+    ADC_ChannelConfig(ADC1, (1 << adc_channel), ADC_SampleTime_239_5Cycles);
+  } else {
+    ADC1->CHSELR &= ~(1 << adc_channel);
+  }
+
+  // Keep internal channels enabled only when set
+  switch (adc_channel) {
+    case ADC_CHANNEL_BAT:
+      ADC_VbatCmd(new_state);
+      break;
+
+    case ADC_CHANNEL_REF:
+      ADC_VrefintCmd(new_state);
+      break;
+
+    case ADC_CHANNEL_TEMP:
+      ADC_TempSensorCmd(new_state);
+      break;
+  }
+
+  s_adc_status.sequence = ADC1->CHSELR;
```

Comment:
Technically I believe this will break if you're enabling/disabling channels during a sequence, but I don't think it really matters for us. Maybe just add a note.